[508] Execute Axe from CMS Preview Server

ncksllvn opened the pull request:
## Update 10/18
Added the `--accessibility` flag to @ncksllvn original full steps below.

## Description
This pull request adds a command line flag, `--accessibility`, that acts as a flag for enabling a new Metalsmith plugin, `injectAxeCheck`. This plugin injects and executes the `axe-core` JS script on each HTML page and processes the result into a banner at the top of the browser. This ultimately allows for us to easily spot accessibility issues both in our content and templates. 

Locally and on regular builds, the `accessibility` flag defaults to false so that it does not disrupt engineers. However, on the CMS Preview Server, it default to true so that content authors can see issues on the page they're previewing.

Ticket - https://github.com/department-of-veterans-affairs/va.gov-team/issues/1794

## Testing done
- Local testing using the `accessibility` flag 
- Local testing using CMS preview server

Full steps:
```
# start your SOCKS proxy!
npm run build -- --pull-drupal --entry=static-pages --accessibility
npm run preview
# navigate to http://localhost:3001/preview?nodeId=67
# Observe yellow banner at top of screen
```

## Screenshots
![image](https://user-images.githubusercontent.com/1915775/66868561-8b5a4980-ef6b-11e9-8ba1-71fbd948bfb6.png)

![image](https://user-images.githubusercontent.com/1915775/66868865-2b17d780-ef6c-11e9-99dd-09ba9d926fd4.png)

## Acceptance criteria
- [x] Visibility into accessibility issues is increased

## Definition of done
- [x] Events are logged appropriately
- [x] Documentation has been updated, if applicable
- [x] A link has been provided to the originating GitHub issue (or connected to it via ZenHub)
- [x] No sensitive information (i.e. PII/credentials/internal URLs/etc.) is captured in logging, hardcoded, or specs

Commit:
[508] Execute Axe from CMS Preview Server

Commit:
Add a comment

Commit:
Add the toggle banner

Commit:
Add a preview build option

Commit:
Also inject onto localhost pages

ncksllvn commented:
Hey @1Copenut, I took a stab at this. The axe-core script and the script to execute it are both injected into the page on the CMS Preview Server and localhost builds (so devs can see markup issues too, in case it's in the template, React widget, or just to increase the visibility of this plugin). The axe execution is scoped to just the `main` element, so we don't see the same site-wide warnings on every page. Want to pull this branch and mess with it? Here's what the rendered banner looks like -

![image](https://user-images.githubusercontent.com/1915775/66784064-62708080-eea7-11e9-83fe-919f12f8c63b.png)

I like the current implementation because it's contained to just that one Metalsmith plugin in this PR. However, it isn't perfect, and I'm wondering how we should improve. 

1. This element is appended at the top of the `header` element. Is that correct, or is the accessibility banner itself breaking a rule? 😆 Also, are there visual issues besides it just being ugly in general?
1. The code I wrote to render the banner is ES6. However, it's injected directly into the browser, so it's not going through our Webpack chain, meaning it's not compiled into old JS, so it's not friendly to old browsers.  Should we convert it to ES5, move it into the Webpack chain (then use React), or is it fine as-is, since it isn't user-facing?

ncksllvn commented:
You can also see it on the Heroku preview server - https://vetsgov-pr-10836.herokuapp.com/disability/eligibility/

Commit:
Merge remote-tracking branch 'origin/master' into 508-preview-server-vagov-1794

ncksllvn commented on src/site/stages/build/plugins/inject-axe-core.js:
```diff
@@ -0,0 +1,125 @@
+
+        const header = document.querySelector('header');
+
+        header.prepend(bannerEl, header.firstChild);
```
This code is ES6 but it's injected directly into the browser, so it isn't compiled down to ES5. So, we either need to convert it to ES5 or move this whole render into the Webpack build, probably as a React component. Or, consider it okay as-is since it isn't user-facing (although CMS users do see it. I wonder if some of them are using IE 🤔 ) I really don't want to move it into React because I like this build step being self-contained.

1Copenut replied:
@ncksllvn This is magical. I'll give it a look tonight after dinner. I'm helping Rian with another IE-specific issue.

I think it'll be worth stepping down to ES5 because there's at least a chance IE11 is still being used for the moment. I'm hopeful Edge will be released on Chromium soon and adopted.

ncksllvn replied:
Oof, just realized the `≤details>` element isn't supported either by IE 🤔

tlei123 committed:
1794 - Ungate Disability Combined Rating Calculator e2e test.

tlei123 commented on src/applications/disability-benefits/disability-rating-calculator/tests/00-all-fields.e2e.spec.js:
```diff
@@ -29,8 +28,6 @@ module.exports = E2eHelpers.createE2eTest(client => {
 
   // Test row deletion.
-  // TODO: Once Design Team has spec'ed out the relevant UX,
-  // cover edge-case for deleting row AFTER calculating combined-rating.
```
Deleting this TODO - No UX design-update forthcoming.

tlei123 committed:
Revert "1794 - Ungate Disability Combined Rating Calculator e2e test."

This reverts commit 0da38f5d4ea1663b883676a968c4f3698c663f35.

Commit:
Convert to ES5

Commit:
Merge branch '508-preview-server-vagov-1794' of https://github.com/department-of-veterans-affairs/vets-website into 508-preview-server-vagov-1794

Commit:
Switch to a command line flag

Commit:
Inject axe-core into body of page to get around assets caching issue

ncksllvn commented:
@1Copenut I changed this to use an explicit build flag to toggle this new build step. So, to run it, do `npm run watch -- --accessibility`. Maybe the argument name could be better? Up to you. This build flag defaults to true for the preview server and false for regular builds/watch, so that it isn't disruptive to engineers. I also converted the JS script to ES5, so it should be okay for review 😄

Commit:
Inject axe-core as separate JS script, instead of body of page

ncksllvn commented on src/site/assets/js/execute-axe-check.js:
```diff
@@ -0,0 +1,78 @@
+        'color-contrast': { enabled: false },
+      },
+    }, processAxeCheckResults);
+})();
```
This has to be ES5 because it's run directly in the browser instead of through Webpack. It isn't public-facing, but it is CMS user-facing since it's executed from the CMS Preview Server. Some of those authors could be old-school, and use IE 😒

1Copenut commented:
I added a few styles on my local machine, take a look and let me know what you think. If these are in keeping, I'll push the change. Will test on the Windows machine in the morning and follow up about IE support.

---
<img width="689" alt="Screen Shot 2019-10-15 at 10 25 23 PM" src="https://user-images.githubusercontent.com/934879/66885734-d50e5880-ef9a-11e9-8fb2-bbce40b16f33.png">

kelsonic commented on src/site/assets/js/execute-axe-check.js:
```diff
@@ -0,0 +1,78 @@
+    console.log('Error executing the Axe check!');
+    console.log(error);
+    return;
+  } else if (results.violations.length === 0) {
```
You could remove the `else` here 👍

kelsonic reviewed (approved):
Nicely done! Looks great! 🎉

1Copenut commented on src/site/assets/js/execute-axe-check.js:
```diff
@@ -0,0 +1,78 @@
+  var details = '<details class="vads-u-background-color--gold-lighter vads-u-padding--1">';
+  details += '<summary>There are (' + results.violations.length + ') accessibility issues on this page.</summary>';
+
+  var violationsList = '<ul class="usa-unstyled-list vads-u-padding-y--2 vads-u-padding-x--6">';
```
Can we add two classes to the UL? ` vads-u-border-color--primary-darker vads-u-border-top--1px`

1Copenut commented on src/site/assets/js/execute-axe-check.js:
```diff
@@ -0,0 +1,78 @@
+  var bannerEl = document.createElement('div');
+
+  var details = '<details class="vads-u-background-color--gold-lighter vads-u-padding--1">';
+  details += '<summary>There are (' + results.violations.length + ') accessibility issues on this page.</summary>';
```
I'd like to add a few classes and an H4 to this block to adjust the background color and offer some more visual contrast:

```javascript
var details = '<details class="vads-u-background-color--primary-alt-lightest  vads-u-padding--1 vads-u-border-color--primary-darker vads-u-border-bottom--2px">';
details += '<summary><h4 class="vads-u-margin-y--2 vads-u-display--inline-block">There are (' + results.violations.length + ') accessibility issues on this page.</h4></summary>';
```

1Copenut commented on src/site/assets/js/execute-axe-check.js:
```diff
@@ -0,0 +1,78 @@
+    var violationEl = '<li class="vads-u-margin-y--1">';
+
+    violationEl += '<details>';
+    violationEl += '<summary>' + violation.help + '</summary>';
```
Can we add a `<strong>` to this line?

```javascript
violationEl += '<summary><strong>' + violation.help + '</strong></summary>';
```

1Copenut reviewed (changes requested):
I have 3 style update requests. I tried these out on Chrome, Safari, and Firefox on MacOS Mojave. Works great. No need for further review on my part after this.

I'll test in Staging on Windows as soon as we're live and I get a URL to do so. Thanks so much @ncksllvn !

Commit:
Adding some CSS classes to error banner.

1Copenut commented:
Blue styling suggestions from [my earlier comment](https://github.com/department-of-veterans-affairs/vets-website/pull/10836#pullrequestreview-302704059) have been added and tested.
